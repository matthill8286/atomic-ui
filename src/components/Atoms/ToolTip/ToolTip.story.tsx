import { boolean, text } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/react'
import * as React from 'react'
import { Link } from '@/components/Atoms/Link'
import { StorybookWrapper } from '@/utils/StorybookWrapper'
import { ToolTip } from './ToolTip'
import { ToolTipProps } from './ToolTip.interface'
import { Icon } from '@/components/Atoms/Icon'
import { OtherInfoOutlined } from '@matthill8286/atomic-icon-library'

const knobs = (primary?: boolean): ToolTipProps => {
  return {
    content: text('Content', 'Lorem Ipsum'),
    primary: boolean('Primary', !!primary),
  }
}

storiesOf('Design System/Atoms/ToolTip', module)
  .add('Default', () => (
    <StorybookWrapper>
      <ToolTip {...knobs()}>
        <Link underline={false}>Link</Link>
      </ToolTip>
    </StorybookWrapper>
  ))
  .add('Primary', () => (
    <StorybookWrapper>
      <ToolTip {...knobs(true)}>
        <Link underline={false}>Link</Link>
      </ToolTip>
    </StorybookWrapper>
  ))
  .add('Icon', () => (
    <StorybookWrapper>
      <ToolTip {...knobs()}>
        <Icon>
          <OtherInfoOutlined height="xs" width="xs" />
        </Icon>
      </ToolTip>
    </StorybookWrapper>
  ))
