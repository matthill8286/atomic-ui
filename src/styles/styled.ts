// eslint-disable-next-line no-restricted-imports
import baseStyled, {
  BaseThemedCssFunction,
  css as baseCss,
  ThemedStyledInterface,
  ThemeProps,
} from 'styled-components'
import { Theme } from '@/types/theme'

export { ThemeConsumer, ThemeContext, keyframes } from 'styled-components'

export type ThemedStyledProps<P, T> = P & ThemeProps<T>
export const styled = baseStyled as ThemedStyledInterface<Theme>
export const css = baseCss as BaseThemedCssFunction<Theme>
export { ThemeProvider } from './ThemeProvider'
