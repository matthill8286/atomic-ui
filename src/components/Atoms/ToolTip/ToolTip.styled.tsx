import { styled } from '@/styles/styled'
import { StyledToolTipContentProps, StyledToolTipProps } from './ToolTip.interface'

export const StyledToolTipWrapper = styled.div`
  position: relative;
  display: inline-block;
  cursor: pointer;
`

export const StyledToolTipContent = styled.div<StyledToolTipContentProps>`
  background: ${({ theme, primary }) => (primary ? theme.color.grey4 : theme.color.white)};
  padding: 10px 0;
  border-radius: ${({ theme }) => theme.dimension.borderRadius2};
`

export const StyledToolTip = styled.div<StyledToolTipProps>`
  border-radius: ${({ theme }) => theme.dimension.borderRadius2};
  box-sizing: border-box;
  display: inline-block;
  position: absolute;
  margin-top: 20px;
  padding: 0 ${({ theme }) => theme.spacing.base.xs};
  z-index: 10;
  left: 50%;
  width: 120px;
  transform: translateX(-50%);
  background: ${({ theme, primary }) => (primary ? theme.color.grey4 : theme.color.white)};
  box-shadow: ${({ theme }) => theme.dimension.elevationLevel2};

  &::before {
    content: '';
    box-shadow: ${({ theme }) => theme.dimension.elevationLevel2};
    background: ${({ theme, primary }) => (primary ? theme.color.grey4 : theme.color.white)};
    display: block;
    position: absolute;
    width: 14px;
    top: -7px;
    z-index: -1;
    left: 50%;
    margin-left: -7px;
    height: 14px;
    transform: rotate(45deg);
  }
`
