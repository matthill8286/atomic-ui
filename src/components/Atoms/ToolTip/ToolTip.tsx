import React, { useState } from 'react'
import { InfoText } from '../Typography'
import { ToolTipProps } from './ToolTip.interface'
import { StyledToolTip, StyledToolTipContent, StyledToolTipWrapper } from './ToolTip.styled'

export const ToolTip: React.FC<ToolTipProps> = props => {
  const { children, content, primary = false, limitLines = 5, ...otherProps } = props
  const [tooltipVisible, setState] = useState(false)

  return (
    <StyledToolTipWrapper
      onMouseEnter={() => setState(true)}
      onMouseLeave={() => setState(false)}
      onClick={() => setState(!tooltipVisible)}
      {...otherProps}>
      {children}
      {tooltipVisible && content && (
        <StyledToolTip primary={primary}>
          <StyledToolTipContent primary={primary}>
            <InfoText
              color={primary ? 'white' : 'grey4'}
              limitLines={limitLines}
              textAlign={'center'}>
              {content}
            </InfoText>
          </StyledToolTipContent>
        </StyledToolTip>
      )}
    </StyledToolTipWrapper>
  )
}
