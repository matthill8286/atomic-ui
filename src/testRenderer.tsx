import { mount, ReactWrapper } from 'enzyme'
import React from 'react'
import { MemoryRouter as Router } from 'react-router-dom'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'
import { saiyanTheme } from '@/styles/sc-vars-saiyan'
import { Theme } from '@/types'

export function renderWithTheme<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>,
  theme = saiyanTheme
) {
  return renderer.create(<ThemeProvider theme={theme}>{tree}</ThemeProvider>).toJSON()
}

export function renderWithThemeAndRouter(tree: React.ReactElement, theme: Theme = saiyanTheme) {
  return renderer
    .create(
      <ThemeProvider theme={theme}>
        <Router>{tree}</Router>
      </ThemeProvider>
    )
    .toJSON()
}

export function renderWithRouter(tree: React.ReactElement) {
  return renderer.create(<Router>{tree}</Router>).toJSON()
}

export function mountWithTheme<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>
): ReactWrapper<P, S, C> {
  return mount(<ThemeProvider theme={saiyanTheme}>{tree}</ThemeProvider>)
}

export function mountWithThemeAndRouter<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>
): ReactWrapper<P, S, C> {
  return mount(
    <ThemeProvider theme={saiyanTheme}>
      <Router>{tree}</Router>
    </ThemeProvider>
  )
}
