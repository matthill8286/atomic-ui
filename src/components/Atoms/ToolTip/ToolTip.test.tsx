import * as React from 'react'
import { mountWithTheme, renderWithThemeAndRouter } from '@/testRenderer'
import { ToolTip } from './ToolTip'
import { StyledToolTipContent, StyledToolTipWrapper } from './ToolTip.styled'

describe('ToolTip', () => {
  it('renders basic', () => {
    const tree = renderWithThemeAndRouter(<ToolTip />)
    expect(tree).toMatchSnapshot()
  })

  it('renders default', () => {
    const comp = mountWithTheme(
      <ToolTip content="Lorem ipsum">
        <span>Link</span>
      </ToolTip>
    )
    expect(comp.find(StyledToolTipWrapper)).toHaveLength(1)
    expect(comp).toMatchSnapshot()
  })

  it('renders primary', () => {
    const comp = mountWithTheme(
      <ToolTip content="Lorem ipsum" primary>
        <span>Link</span>
      </ToolTip>
    )
    expect(comp.find(StyledToolTipWrapper)).toHaveLength(1)
    expect(comp).toMatchSnapshot()
  })

  it('renders visible on hover and hides on leave', () => {
    const comp = mountWithTheme(
      <ToolTip content="Lorem ipsum">
        <span>Link</span>
      </ToolTip>
    )
    expect(comp.find(StyledToolTipContent)).toHaveLength(0)

    comp
      .find(StyledToolTipWrapper)
      .first()
      .simulate('mouseenter')
    expect(comp.find(StyledToolTipContent)).toHaveLength(1)

    comp
      .find(StyledToolTipWrapper)
      .first()
      .simulate('mouseleave')
    expect(comp.find(StyledToolTipContent)).toHaveLength(0)
  })
})
