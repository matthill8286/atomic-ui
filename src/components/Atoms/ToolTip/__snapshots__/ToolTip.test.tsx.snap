// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToolTip renders basic 1`] = `
.c0 {
  position: relative;
  display: inline-block;
  cursor: pointer;
}

<div
  className="c0"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
/>
`;

exports[`ToolTip renders default 1`] = `ReactWrapper {}`;

exports[`ToolTip renders primary 1`] = `ReactWrapper {}`;
